incubator/hnc/internal/validators/hierarchy.go

```diff
@@ -210,9 +212,54 @@ func (v *Hierarchy) checkParent(ns, curParent, newParent *forest.Namespace) admi
 		return deny(metav1.StatusReasonConflict, "Illegal parent: "+reason)
 	}
 
+	// Prevent overwriting source objects in the descendants after the hierarchy change.
+	if co := v.hasConflictObject(newParent, ns); len(co) != 0 {
+		msg := "Cannot update hierarchy because it would overwrite the following object(s):\n"
+		msg += "  * " + strings.Join(co, "\n  * ") + "\n"
+		msg += "To fix this, please rename or remove the conflicting objects first."
+		return deny(metav1.StatusReasonConflict, msg)
+	}
+
 	return allow("")
 }
 
+// hasConflictObject returns a list of namespaced objects if there's any conflict.
+func (v *Hierarchy) hasConflictObject(pns, cns *forest.Namespace) []string {
```

Comment:
Two nits:
* "has" suggests a boolean, but this returns a slice. `getConflictingObjects` would also match up well with `getConflictingObjectsOfType`, below.
* Instead of `pns` and `cns`, I'd keep the `newParent` and `ns` variable names from the caller, both for consistency but also because IMO it more clearly shows which is the "main" namespace (`ns`/`cns`) and what's the thing that's being tested against it.

Reply:
Note that `pns/cns` is used in this function and also `getConflictingObjectsOfType`

Reply:
sgtm, updated!